eng/common/TestResources/Remove-TestResources.ps1

```diff
@@ -213,16 +213,38 @@ $verifyDeleteScript = {
     }
 }
 
+# Get any Key Vaults that will be deleted so they can be purged later if soft delete is enabled.
+$vaults = Get-AzKeyVault -ResourceGroupName "$ResourceGroupName" | ForEach-Object {
```

Comment:
We should figure out a way where we can share this code for purging specific resources like this with this script and our resource clean-up script (https://github.com/Azure/azure-sdk-tools/blob/main/eng/scripts/live-test-resource-cleanup.ps1) that we run. Perhaps just a helper ps1 we can use.

Reply:
I think it may evolve into some sort of deletion handler library for resource types. Keyvault and storage would be included right now, but I imagine we'll encounter more in the future.

Reply:
I could, but it will still have to live in eng/common so any scripts outside that will have to import it. But I think the discovery and purging wouldn't necessarily work in all cases. Here, I need to limit it to the resource group. I could make that optional in this case, but how much longer till we start bifurcating business logic to tailor for different uses? For now, at least, these blocks aren't doing all that much that is outside of normal cmdlet usage. I think we should cross that bridge when we get there.

Reply:
Yes I would expect it to be under eng/common still. It really only has to be a discover and clean-up function that takes in a resource group so that we keep the logic consistent between both clean-up mechanisms.

Reply:
It has to be two functions. Can't discover resources from a resource group that was just deleted. For general clean-up, you should be caring about resource groups - if something, like a Key Vault or *especially* Managed HSM is in a deleted state (perhaps for X days), purge it. So already there's a difference: this has to check for resources in a resource group, and clean-up does not.

Again, it's not that hard to make the resource group optional, but already business logic is separate when the rest of the clean-up is common. I feel we're making this more complicated from the start for something that may not be necessary or even worth it.

Reply:
I'm fine waiting to see what the other logic looks like and then deciding. Did you plan to add that to the clean-up script as well or were you expecting one of us to port that over?

Today the clean-up script only attempts to clean-up things inside a resource group with an expired DeleteAfter tag so I was expecting we would likely only do this purging in those groups as well but you make a good point that maybe we should purge deleted things no matter what group they are in.

Reply:
I can. I did refactor these out into another script I'm testing now with TestResources.

We should more aggressively clean up deleted Managed HSMs, but there's no PowerShell support for that yet and it should be isolated to only sdk/keyvault for now, which we'll handle with -post scripts.

Reply:
On a related note, quite a few Key Vaults sitting in a deleted state, not that really cost us anything apart from pressure on our quota:

![image](https://user-images.githubusercontent.com/1532486/129250717-fd5a4ed7-8870-41d7-8d1e-2364ff88f560.png)

Deleted Managed HSMs cost the same as active ones. There are a couple, including one apparently using the default 90d retention policy. I'll attempt to purge them. We really need to get #1893 implemented.